packages/block-library/src/table-of-contents/utils.js

```diff
@@ -0,0 +1,123 @@
+
+/**
+ * External dependencies
+ */
+import slugify from 'slugify';
+
+/**
+ * WordPress dependencies
+ */
+import { select } from '@wordpress/data';
+
+export function linearToNestedList( array, index = 0 ) {
+	const returnValue = [];
+
+	array.forEach( function( heading, key ) {
+		if ( typeof heading.content === 'undefined' ) {
+			return;
+		}
+
+		// Make sure we are only working with the same level as the first iteration in our set.
+		if ( heading.level === array[ 0 ].level ) {
+			// Check that the next iteration will return a value.
+			// If it does and the next level is greater than the current level,
+			// the next iteration becomes a child of the current interation.
+			if (
+				( typeof array[ key + 1 ] !== 'undefined' ) &&
+				( array[ key + 1 ].level > heading.level )
+			) {
+				// We need to calculate the last index before the next iteration that has the same level (siblings).
+				// We then use this last index to slice the array for use in recursion.
+				// This prevents duplicate nodes.
+				let endOfSlice = array.length;
+				for ( let i = ( key + 1 ); i < array.length; i++ ) {
+					if ( array[ i ].level === heading.level ) {
+						endOfSlice = i;
+						break;
+					}
+				}
+
+				// We found a child node: Push a new node onto the return array with children.
+				returnValue.push( {
+					block: heading,
+					index: index + key,
+					children: linearToNestedList( array.slice( key + 1, endOfSlice ), index + key + 1 ),
+				} );
+			} else {
+				// No child node: Push a new node onto the return array.
+				returnValue.push( {
+					block: heading,
+					index: index + key,
+					children: null,
+				} );
+			}
+		}
+	} );
+
+	return returnValue;
+}
+
+export function getPageHeadings() {
+	return convertHeadingBlocksToAttributes( getHeadingBlocks() );
+}
+
+export function getHeadingBlocks() {
+	const editor = select( 'core/editor' );
+	return editor.getBlocks().filter( ( block ) => block.name === 'core/heading' );
+}
+
+export function convertHeadingBlocksToAttributes( headingBlocks ) {
+	return headingBlocks.map( function( heading ) {
+		const level = heading.attributes.level.toString();
+
+		const headingContent = heading.attributes.content || '';
+		const anchorContent = heading.attributes.anchor || '';
+
+		// strip html from heading and attribute content
+		const contentDiv = document.createElement( 'div' );
+
+		contentDiv.innerHTML = headingContent;
+		const content = contentDiv.textContent || contentDiv.innerText || '';
+
+		contentDiv.innerHTML = anchorContent;
+		let anchor = contentDiv.textContent || contentDiv.innerText || '';
+
+		if ( anchor !== '' && anchor.indexOf( '#' ) === -1 ) {
+			anchor = '#' + slugify( anchor, { remove: /[^\w\s-]/g } );
+		}
+
+		return { content, anchor, level };
+	} );
+}
+
+export function updateHeadingBlockAnchors() {
+	// Add anchors to any headings that don't have one.
```

Comment:
Should we just put this logic to the heading blocks directly? Instead of modifying them from the outside.

Reply:
That could be done. This was carried over from the plugin this was based on. It however is a requirement for this block alone that may not be required/wanted by the heading blocks. The other option i thought of having here was not to alter the heading id's at all but just link the ones that do have an anchor. Not sure which way to go but Id however love suggestions on the best way to implement this linking.